### PR TITLE
[Inference/Feat] Feat quant kvcache step2

### PR DESCRIPTION
**Pull request by Courtesy-Xs**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [ ] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?
This PR add quant kvcache support for context_kv_cache_memcpy
> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Comment by Courtesy-Xs** [comment]
image

File: extensions/csrc/kernel/cuda/utils/vec_copy.h [hunk 1]
```diff
@@ -33,9 +32,33 @@ __device__ __inline__ void copy_zero_vector(T *dst) {
   *(reinterpret_cast<VT *>(dst)) = funcs::CastFunctor<float, VT>()(0.0f);
 }
 
+template <typename SrcT, typename DstT, int vec_size>
+__device__ __inline__ void copy(const SrcT *src, DstT *dst) {
+  using SrcVT = typename common::VecTypeTrait<SrcT, vec_size>::Type;
+  using DstVT = typename common::VecTypeTrait<DstT, vec_size>::Type;
+  // Note(LiuYang): Here static_cast can't be used for cast between two pointer
+  *(reinterpret_cast<DstVT *>(dst)) = funcs::CastFunctor<SrcVT, DstVT>()(
```
Comment (yuanheng-zhao) [inline extensions/csrc/kernel/cuda/utils/vec_copy.h:40]: Wonder if `static_cast` here in `CastFunctor` was accidentally added, or it (only) aims at compile-time check?

File: extensions/csrc/kernel/cuda/context_kv_cache_memcpy_kernel.cu [hunk 2]
```diff
@@ -161,26 +162,63 @@ void apply_context_kv_cache_memcpy(
...
-    sequence_lengths,
-    cu_seqlens,
-    block_tables,
-    max_seq_len_in_batch
-    );)
+
+    TORCH_CHECK(key.scalar_type() == at::ScalarType::Float || key.scalar_type() == at::ScalarType::Half || key.scalar_type() == at::ScalarType::BFloat16,
+    "Dtype of key should be float, half or bfloat16!");
+    TORCH_CHECK(key_cache.scalar_type() == at::ScalarType::Byte || key_cache.scalar_type() == key.scalar_type(),
+    "Dtype of query and kvcache should be the same unless dtype of kvcache is fp8!");
```
Comment (yuanheng-zhao) [inline extensions/csrc/kernel/cuda/context_kv_cache_memcpy_kernel.cu:178]: trivial typo

**Review by yuanheng-zhao — APPROVED** [review]
LGTM

File: extensions/csrc/kernel/cuda/context_kv_cache_memcpy_kernel.cu (hunk 2 above)
Comment (Courtesy-Xs) [reply extensions/csrc/kernel/cuda/context_kv_cache_memcpy_kernel.cu:178]: Next PR I will fix it

File: extensions/csrc/kernel/cuda/utils/vec_copy.h (hunk 1 above)
Comment (Courtesy-Xs) [reply extensions/csrc/kernel/cuda/utils/vec_copy.h:40]: Next PR I will delete this note

**Pull request merged**